File: src/utils/cloudDocuments.ts
import type { Connection } from '../types/circuit'
import {
  PERSISTED_BOARD_FORMAT,
  PERSISTED_BOARD_VERSION,
  type PersistedBoard,
} from '../types/persistence'
import type { Relation, Widget, WidgetGlue } from '../types/spatial'

export const CLOUD_INDEX_FORMAT = 'grovepad-board-index' as const
const CLOUD_INDEX_VERSION = 1 as const
export const CLOUD_CANVAS_FORMAT = 'grovepad-canvas' as const
export const CLOUD_CANVAS_VERSION = 1 as const

export interface CloudBoardIndexDocument {
  format: typeof CLOUD_INDEX_FORMAT
  v: typeof CLOUD_INDEX_VERSION
  boardFormat: typeof PERSISTED_BOARD_FORMAT
  boardVersion: typeof PERSISTED_BOARD_VERSION
  workspaces: PersistedBoard['workspaces']
  canvases: PersistedBoard['canvases']
  activePacks: PersistedBoard['activePacks']
  relations: PersistedBoard['relations']
  connections: PersistedBoard['connections']
  glues: PersistedBoard['glues']
  extra: Record<string, unknown>
}

export interface CloudCanvasDocument {
  format: typeof CLOUD_CANVAS_FORMAT
  v: typeof CLOUD_CANVAS_VERSION
  canvasId: string
  widgets: PersistedBoard['widgets']
  relations: PersistedBoard['relations']
  connections: PersistedBoard['connections']
  glues: PersistedBoard['glues']
}

export interface SplitCloudBoard {
  index: CloudBoardIndexDocument
  canvases: Record<string, CloudCanvasDocument>
}

export interface EncodedCloudDocument {
  body: string
  encoding: 'gzip' | 'identity'
  checksum: string
  byteLength: number
  uncompressedBytes: number
}

const CORE_BOARD_FIELDS = new Set([
  'format',
  'v',
  'workspaces',
  'canvases',
  'widgets',
  'relations',
  'connections',
  'glues',
  // Legacy grouping records — dropped at the split boundary.
  'groups',
  'activePacks',
  // Legacy device fields are intentionally discarded at the split boundary.
  'activeWorkspaceId',
  'activeCanvasId',
  'canvasViews',
])

function endpointCanvas(
  widgets: Record<string, Widget>,
  fromId: string,
  toId: string,
): string | null {
  const fromCanvas = widgets[fromId]?.canvasId
  return fromCanvas && fromCanvas === widgets[toId]?.canvasId ? fromCanvas : null
}

function glueCanvas(widgets: Record<string, Widget>, glue: WidgetGlue): string | null {
  const first = widgets[glue.widgetIds[0]!]?.canvasId
  if (!first) return null
  return glue.widgetIds.every((widgetId) => widgets[widgetId]?.canvasId === first)
    ? first
    : null
}

/** Split one canonical board into a small index and independently writable canvases. */
export function splitCloudBoard(board: PersistedBoard): SplitCloudBoard {
  const canvases = Object.fromEntries(
    Object.keys(board.canvases).map((canvasId) => [canvasId, {
      format: CLOUD_CANVAS_FORMAT,
      v: CLOUD_CANVAS_VERSION,
      canvasId,
      widgets: {},
      relations: {},
      connections: {},
      glues: {},
    } satisfies CloudCanvasDocument]),
  ) as Record<string, CloudCanvasDocument>
  for (const [widgetId, widget] of Object.entries(board.widgets)) {
    const canvas = canvases[widget.canvasId]
    if (canvas) canvas.widgets[widgetId] = widget
  }

  const indexRelations: Record<string, Relation> = {}
  for (const [relationId, relation] of Object.entries(board.relations)) {
    const canvasId = endpointCanvas(board.widgets, relation.fromId, relation.toId)
    const canvas = canvasId ? canvases[canvasId] : undefined
    if (canvas) canvas.relations[relationId] = relation
    else indexRelations[relationId] = relation
  }

  const indexConnections: Record<string, Connection> = {}
  for (const [connectionId, connection] of Object.entries(board.connections)) {
    const canvasId = endpointCanvas(board.widgets, connection.fromId, connection.toId)
    const canvas = canvasId ? canvases[canvasId] : undefined
    if (canvas) canvas.connections[connectionId] = connection
    else indexConnections[connectionId] = connection
  }

  const indexGlues: Record<string, WidgetGlue> = {}
  for (const [glueId, glue] of Object.entries(board.glues)) {
    const canvasId = glueCanvas(board.widgets, glue)
    const canvas = canvasId ? canvases[canvasId] : undefined
    if (canvas) canvas.glues[glueId] = glue
    else indexGlues[glueId] = glue
  }

  const extra = Object.fromEntries(
    Object.entries(board).filter(([key]) => !CORE_BOARD_FIELDS.has(key)),
  )
  return {
    index: {
      format: CLOUD_INDEX_FORMAT,
      v: CLOUD_INDEX_VERSION,
      boardFormat: PERSISTED_BOARD_FORMAT,
      boardVersion: PERSISTED_BOARD_VERSION,
      workspaces: board.workspaces,
      canvases: board.canvases,
      activePacks: board.activePacks,
      relations: indexRelations,
      connections: indexConnections,
      glues: indexGlues,
      extra,
    },
    canvases,
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

export function isCloudBoardIndex(value: unknown): value is CloudBoardIndexDocument {
  return isRecord(value) &&
    value.format === CLOUD_INDEX_FORMAT &&
    value.v === CLOUD_INDEX_VERSION &&
    value.boardFormat === PERSISTED_BOARD_FORMAT &&
    value.boardVersion === PERSISTED_BOARD_VERSION &&
    isRecord(value.workspaces) &&
    isRecord(value.canvases) &&
    Array.isArray(value.activePacks) &&
    isRecord(value.relations) &&
    isRecord(value.connections) &&
    // Accept legacy documents that carry `groups` instead — their grouping
    // records are ignored on join, but the document itself remains readable.
    (isRecord(value.glues) || isRecord(value.groups)) &&
    isRecord(value.extra)
}

export function isCloudCanvasDocument(value: unknown): value is CloudCanvasDocument {
  return isRecord(value) &&
    value.format === CLOUD_CANVAS_FORMAT &&
    value.v === CLOUD_CANVAS_VERSION &&
    typeof value.canvasId === 'string' &&
    isRecord(value.widgets) &&
    isRecord(value.relations) &&
    isRecord(value.connections) &&
    (isRecord(value.glues) || isRecord(value.groups))
}

/** Reassemble transport documents; the board parser performs domain validation next. */
export function joinCloudBoard(
  index: CloudBoardIndexDocument,
  canvasDocuments: readonly CloudCanvasDocument[],
): PersistedBoard {
  const widgets: PersistedBoard['widgets'] = {}
  const relations: PersistedBoard['relations'] = { ...index.relations }
  const connections: PersistedBoard['connections'] = { ...index.connections }
  const glues: PersistedBoard['glues'] = { ...(index.glues ?? {}) }
  for (const canvas of canvasDocuments) {
    if (!index.canvases[canvas.canvasId]) continue
    Object.assign(widgets, canvas.widgets)
    Object.assign(relations, canvas.relations)
    Object.assign(connections, canvas.connections)
    Object.assign(glues, canvas.glues ?? {})
  }
  return {
    ...index.extra,
    format: index.boardFormat,
    v: index.boardVersion,
    workspaces: index.workspaces,
    canvases: index.canvases,
    widgets,
    relations,
    connections,
    glues,
    activePacks: index.activePacks,
  }
}

function canonicalizeJsonValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalizeJsonValue)
  if (!isRecord(value)) return value
  return Object.fromEntries(
    Object.keys(value)
      .sort()
      .map((key) => [key, canonicalizeJsonValue(value[key])]),
  )
}

/** Stable JSON bytes: object insertion order never creates a false change. */
export function canonicalJson(value: unknown): string {
  return JSON.stringify(canonicalizeJsonValue(value))
}

export async function sha256Hex(value: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(value))
  return [...new Uint8Array(digest)]
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('')
}

function bytesToBytea(bytes: Uint8Array): string {
  return `\\x${[...bytes].map((byte) => byte.toString(16).padStart(2, '0')).join('')}`
}

function byteaToBytes(value: string): Uint8Array {
  const hex = value.startsWith('\\x') ? value.slice(2) : value
  if (hex.length % 2 !== 0 || !/^[\da-f]*$/i.test(hex)) throw new Error('Invalid bytea body')
  return Uint8Array.from(hex.match(/.{2}/g) ?? [], (pair) => Number.parseInt(pair, 16))
}

async function transformBytes(
  bytes: Uint8Array,
  format: 'gzip' | 'identity',
  direction: 'compress' | 'decompress',
): Promise<Uint8Array> {
  if (format === 'identity') return bytes
  const stream = direction === 'compress'
    ? new CompressionStream('gzip')
    : new DecompressionStream('gzip')
  const source = new Blob([bytes as BlobPart]).stream().pipeThrough(stream)
  return new Uint8Array(await new Response(source).arrayBuffer())
}

export async function encodeCloudDocument(value: unknown): Promise<EncodedCloudDocument> {
  const json = canonicalJson(value)
  const source = new TextEncoder().encode(json)
  const supportsGzip = typeof CompressionStream !== 'undefined'
  const encoding = supportsGzip ? 'gzip' : 'identity'
  const body = await transformBytes(source, encoding, 'compress')
  return {
    body: bytesToBytea(body),
    encoding,
    checksum: await sha256Hex(json),
    byteLength: body.byteLength,
    uncompressedBytes: source.byteLength,
  }
}

export async function decodeCloudDocument(
  body: string,
  encoding: 'gzip' | 'identity',
  checksum: string,
): Promise<unknown> {
  const compressed = byteaToBytes(body)
  const bytes = await transformBytes(compressed, encoding, 'decompress')
  const json = new TextDecoder().decode(bytes)
  if (await sha256Hex(json) !== checksum) throw new Error('Cloud document checksum mismatch')
  return JSON.parse(json) as unknown
}

File: apple/Sources/GrovepadCore/Persistence/CloudDocuments.swift
import Foundation

// ---------------------------------------------------------------------------
// Port of `src/utils/cloudDocuments.ts`: one small board index plus one
// document per canvas, the canonical JSON form checksums are taken over, and
// the join that rebuilds a board document for the parser.
//
// `encodeCloudDocument`/`decodeCloudDocument` (gzip + bytea framing for the
// cloud transport) are not ported here: gzip needs a platform library and
// belongs with the sync client (roadmap phase 6).
// ---------------------------------------------------------------------------

public enum CloudDocuments {
    public static let indexFormat = "grovepad-board-index"
    public static let indexVersion = 1
    public static let canvasFormat = "grovepad-canvas"
    public static let canvasVersion = 1

    /// Fields that belong to the board itself; everything else is `extra`.
    /// Device fields an old payload still echoes are discarded at the split.
    static let coreBoardFields: Set<String> = [
        "format", "v", "workspaces", "canvases", "widgets", "relations", "connections", "glues",
        "groups", "activePacks", "activeWorkspaceId", "activeCanvasId", "canvasViews",
    ]

    public struct Split: Equatable {
        public var index: JSONObject
        public var canvases: OrderedMap<JSONObject>

        public init(index: JSONObject, canvases: OrderedMap<JSONObject>) {
            self.index = index
            self.canvases = canvases
        }
    }

    // MARK: - Split

    /// `splitCloudBoard(board)` over a serialized board document.
    public static func splitCloudBoard(_ board: JSONObject) -> Split {
        let widgets = board.object("widgets") ?? JSONObject()
        var canvases = OrderedMap<JSONObject>()
        for canvasId in (board.object("canvases") ?? JSONObject()).keys {
            var document = JSONObject()
            document["format"] = .string(canvasFormat)
            document["v"] = .number(Double(canvasVersion))
            document["canvasId"] = .string(canvasId)
            document["widgets"] = .object(JSONObject())
            document["relations"] = .object(JSONObject())
            document["connections"] = .object(JSONObject())
            document["glues"] = .object(JSONObject())
            canvases[canvasId] = document
        }

        func place(_ canvasId: String?, _ key: String, _ id: String, _ value: JSONValue) -> Bool {
            guard let canvasId, var canvas = canvases[canvasId] else { return false }
            var bucket = canvas.object(key) ?? JSONObject()
            bucket[id] = value
            canvas[key] = .object(bucket)
            canvases[canvasId] = canvas
            return true
        }

        for (widgetId, widget) in widgets.entries {
            _ = place(JS.key(widget["canvasId"]), "widgets", widgetId, widget)
        }

        var indexRelations = JSONObject()
        for (relationId, relation) in (board.object("relations") ?? JSONObject()).entries {
            let canvasId = endpointCanvas(widgets, JS.key(relation["fromId"]), JS.key(relation["toId"]))
            if !place(canvasId, "relations", relationId, relation) { indexRelations[relationId] = relation }
        }

        var indexConnections = JSONObject()
        for (connectionId, connection) in (board.object("connections") ?? JSONObject()).entries {
            let canvasId = endpointCanvas(widgets, JS.key(connection["fromId"]), JS.key(connection["toId"]))
            if !place(canvasId, "connections", connectionId, connection) { indexConnections[connectionId] = connection }
        }

        var indexGlues = JSONObject()
        for (glueId, glue) in (board.object("glues") ?? JSONObject()).entries {
            let canvasId = glueCanvas(widgets, glue)
            if !place(canvasId, "glues", glueId, glue) { indexGlues[glueId] = glue }
        }

        let extra = board.filter { key, _ in !coreBoardFields.contains(key) }
        var index = JSONObject()
        index["format"] = .string(indexFormat)
        index["v"] = .number(Double(indexVersion))
        index["boardFormat"] = .string(BoardFormat.format)
        index["boardVersion"] = .number(Double(BoardFormat.version))
        index["workspaces"] = board["workspaces"] ?? .object(JSONObject())
        index["canvases"] = board["canvases"] ?? .object(JSONObject())
        index["activePacks"] = board["activePacks"] ?? .array([])
        index["relations"] = .object(indexRelations)
        index["connections"] = .object(indexConnections)
        index["glues"] = .object(indexGlues)
        index["extra"] = .object(extra)
        return Split(index: index, canvases: canvases)
    }

    /// The canvas both endpoints sit on, or nil for a cross-canvas edge.
    /// `fromCanvas && fromCanvas === …`: an empty canvas id is falsy.
    private static func endpointCanvas(_ widgets: JSONObject, _ fromId: String?, _ toId: String?) -> String? {
        guard let fromId, let fromCanvas = JS.key(widgets.object(fromId)?["canvasId"]), !fromCanvas.isEmpty else { return nil }
        guard let toId, JS.key(widgets.object(toId)?["canvasId"]) == fromCanvas else { return nil }
        return fromCanvas
    }

    private static func glueCanvas(_ widgets: JSONObject, _ glue: JSONValue) -> String? {
        let widgetIds = glue["widgetIds"]?.arrayValue ?? []
        guard let firstId = widgetIds.first.flatMap(JS.key), let first = JS.key(widgets.object(firstId)?["canvasId"]), !first.isEmpty else { return nil }
        let allOnFirst = widgetIds.allSatisfy { id in
            guard let id = JS.key(id) else { return false }
            return JS.key(widgets.object(id)?["canvasId"]) == first
        }
        return allOnFirst ? first : nil
    }

    // MARK: - Shape checks

    public static func isCloudBoardIndex(_ value: JSONValue) -> Bool {
        guard let object = value.objectValue else { return false }
        return object["format"] == .string(indexFormat) &&
            object["v"] == .number(Double(indexVersion)) &&
            object["boardFormat"] == .string(BoardFormat.format) &&
            object["boardVersion"] == .number(Double(BoardFormat.version)) &&
            object.object("workspaces") != nil &&
            object.object("canvases") != nil &&
            object.array("activePacks") != nil &&
            object.object("relations") != nil &&
            object.object("connections") != nil &&
            // Legacy documents carry `groups` instead of `glues`; still readable.
            (object.object("glues") != nil || object.object("groups") != nil) &&
            object.object("extra") != nil
    }

    public static func isCloudCanvasDocument(_ value: JSONValue) -> Bool {
        guard let object = value.objectValue else { return false }
        return object["format"] == .string(canvasFormat) &&
            object["v"] == .number(Double(canvasVersion)) &&
            JS.isString(object["canvasId"]) &&
            object.object("widgets") != nil &&
            object.object("relations") != nil &&
            object.object("connections") != nil &&
            (object.object("glues") != nil || object.object("groups") != nil)
    }

    // MARK: - Join

    /// Reassemble transport documents; the board parser validates the result.
    public static func joinCloudBoard(index: JSONObject, canvases canvasDocuments: [JSONObject]) -> JSONObject {
        let indexCanvases = index.object("canvases") ?? JSONObject()
        var widgets = JSONObject()
        var relations = index.object("relations") ?? JSONObject()
        var connections = index.object("connections") ?? JSONObject()
        var glues = index.object("glues") ?? JSONObject()
        for canvas in canvasDocuments {
            guard let canvasId = JS.key(canvas["canvasId"]), indexCanvases.contains(canvasId) else { continue }
            widgets.merge(canvas.object("widgets") ?? JSONObject())
            relations.merge(canvas.object("relations") ?? JSONObject())
            connections.merge(canvas.object("connections") ?? JSONObject())
            glues.merge(canvas.object("glues") ?? JSONObject())
        }
        var board = index.object("extra") ?? JSONObject()
        board["format"] = index["boardFormat"] ?? .null
        board["v"] = index["boardVersion"] ?? .null
        board["workspaces"] = index["workspaces"] ?? .null
        board["canvases"] = index["canvases"] ?? .null
        board["widgets"] = .object(widgets)
        board["relations"] = .object(relations)
        board["connections"] = .object(connections)
        board["glues"] = .object(glues)
        board["activePacks"] = index["activePacks"] ?? .null
        return board
    }

    // MARK: - Checksums

    /// Stable JSON bytes: keys sorted at every level, then `JSON.stringify`.
    public static func canonicalJson(_ value: JSONValue) -> String {
        JSONWriter.canonical(value)
    }

    public static func sha256Hex(_ text: String) -> String {
        SHA256.hex(text)
    }

    /// `fingerprintBoard` from syncBaseline.ts: the checksums `fetchCloudHead`
    /// compares before any document crosses the network.
    public struct BoardFingerprint: Equatable {
        public var indexChecksum: String
        public var canvasChecksums: OrderedMap<String>

        public init(indexChecksum: String, canvasChecksums: OrderedMap<String>) {
            self.indexChecksum = indexChecksum
            self.canvasChecksums = canvasChecksums
        }

        /// `fingerprintsMatch`.
        public func matches(_ other: BoardFingerprint) -> Bool {
            guard indexChecksum == other.indexChecksum, canvasChecksums.count == other.canvasChecksums.count else { return false }
            return canvasChecksums.entries.allSatisfy { other.canvasChecksums[$0.key] == $0.value }
        }
    }

    public static func fingerprintBoard(_ document: JSONObject) -> BoardFingerprint {
        let split = splitCloudBoard(document)
        var canvases = OrderedMap<String>()
        for (canvasId, canvas) in split.canvases.entries {
            canvases[canvasId] = sha256Hex(canonicalJson(.object(canvas)))
        }
        return BoardFingerprint(indexChecksum: sha256Hex(canonicalJson(.object(split.index))), canvasChecksums: canvases)
    }

    public static func fingerprintBoard(_ board: Board) -> BoardFingerprint {
        fingerprintBoard(BoardSerializer.serializePersistedBoard(board))
    }
}
